src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -0,0 +1,658 @@
+  T* request_p = (T*)(cast_from_oop<address>(requested_obj) + field_offset);
+  ResourceBitMap* oopmap;
+  address requested_region_bottom;
+
+  if (request_p >= (T*)_requested_closed_region_bottom) {
+    assert(request_p < (T*)_requested_closed_region_top, "sanity");
+    oopmap = _closed_oopmap;
+    requested_region_bottom = _requested_closed_region_bottom;
+  } else {
+    assert(request_p >= (T*)_requested_open_region_bottom, "sanity");
+    assert(request_p <  (T*)_requested_open_region_top, "sanity");
+    oopmap = _open_oopmap;
+    requested_region_bottom = _requested_open_region_bottom;
+  }
+
+  // Mark the pointer in the oopmap
+  T* region_bottom = (T*)requested_region_bottom;
+  assert(request_p >= region_bottom, "must be");
+  BitMap::idx_t idx = request_p - region_bottom;
+  assert(idx < oopmap->size(), "overflow");
+  oopmap->set_bit(idx);
+}
+
+void ArchiveHeapWriter::update_header_for_requested_obj(oop requested_obj, oop src_obj,  Klass* src_klass) {
+  assert(UseCompressedClassPointers, "Archived heap only supported for compressed klasses");
+  narrowKlass nk = ArchiveBuilder::current()->get_requested_narrow_klass(src_klass);
+  address buffered_addr = requested_addr_to_buffered_addr(cast_from_oop<address>(requested_obj));
+
+  oop fake_oop = cast_to_oop(buffered_addr);
+  fake_oop->set_narrow_klass(nk);
+
+  // We need to retain the identity_hash, because it may have been used by some hashtables
+  // in the shared heap. This also has the side effect of pre-initializing the
+  // identity_hash for all shared objects, so they are less likely to be written
+  // into during run time, increasing the potential of memory sharing.
+  if (src_obj != nullptr) {
+    int src_hash = src_obj->identity_hash();
+    fake_oop->set_mark(markWord::prototype().copy_set_hash(src_hash));
+    assert(fake_oop->mark().is_unlocked(), "sanity");
+
+    DEBUG_ONLY(int archived_hash = fake_oop->identity_hash());
+    assert(src_hash == archived_hash, "Different hash codes: original %x, archived %x", src_hash, archived_hash);
+  }
+}
+
+// Relocate an element in the buffered copy of HeapShared::roots()
+template <typename T> void ArchiveHeapWriter::relocate_root_at(oop requested_roots, int index) {
+  size_t offset = (size_t)((objArrayOop)requested_roots)->obj_at_offset<T>(index);
+  relocate_field_in_requested_obj<T>(requested_roots, offset);
+}
+
+class ArchiveHeapWriter::EmbeddedOopRelocator: public BasicOopIterateClosure {
+  oop _src_obj;
+  oop _requested_obj;
+public:
+  EmbeddedOopRelocator(oop src_obj, oop requested_obj) :
+    _src_obj(src_obj), _requested_obj(requested_obj) {}
+
+  void do_oop(narrowOop *p) { EmbeddedOopRelocator::do_oop_work(p); }
+  void do_oop(      oop *p) { EmbeddedOopRelocator::do_oop_work(p); }
+
+private:
+  template <class T> void do_oop_work(T *p) {
+    size_t field_offset = pointer_delta(p, _src_obj, sizeof(char));
+    ArchiveHeapWriter::relocate_field_in_requested_obj<T>(_requested_obj, field_offset);
```

Comment:
I understand what we are doing here, but it seems to be taking a bit of a convoluted approach. It performs quite a few operations to calculate `source_referent`, which is already available in *p.
So an alternate implementation for `do_oop(narrowOop *p)` can be:
```
    if (!CompressedOops::is_null(*p)) {
      size_t field_offset = pointer_delta(p, _src_obj, sizeof(char));
      oop request_referrent = source_obj_to_requested_obj(CompressedOops::decode(*p));  // source referent is already in *p
      narrowOop* field_address_in_buffer = (narrowOop*)(cast_from_oop<address>(_buffered_obj) + field_offset);
      *field_address_in_buffer = CompressedOops::encode(request_referrent);
      mark_oop_pointer<narrowOop>(_requested_obj, field_offset, _oopmap);
    }
```
The `_buffered_obj` can be calculated in the `iterator` below and passed to `EmbeddedOopRelocator` c'tor.
`do_oop(oop *p)` would follow the same. Does this read better?

Reply:
Thanks for the suggestion. I simplified the relocation code by loading/storing the field  using the "buffered address" instead of the "requested address".